File: example/java/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/example/model/input/RecursiveTreeInput.java
package org.babyfish.jimmer.sql.example.model.input;

import lombok.Data;
import org.babyfish.jimmer.Input;
import org.babyfish.jimmer.sql.example.model.TreeNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.mapstruct.BeanMapping;
import org.mapstruct.Mapper;
import org.mapstruct.ReportingPolicy;
import org.mapstruct.factory.Mappers;

import java.util.List;

@Data
public class RecursiveTreeInput implements Input<TreeNode> {

    private final static Converter CONVERTER = Mappers.getMapper(Converter.class);

    private String name;

    @Nullable
    private List<RecursiveTreeInput> childNodes;

    @Override
    public TreeNode toEntity() {
        return CONVERTER.toTreeNode(this);
    }

    @Mapper
    interface Converter {

        @BeanMapping(unmappedTargetPolicy = ReportingPolicy.IGNORE)
        TreeNode toTreeNode(RecursiveTreeInput input);
    }
}

File: example/kotlin/jimmer-sql-kt/src/main/kotlin/org/babyfish/jimmer/sql/example/model/input/RecursiveTreeInput.kt
package org.babyfish.jimmer.sql.example.model.input

import org.babyfish.jimmer.Input
import org.babyfish.jimmer.sql.example.model.TreeNode
import org.mapstruct.BeanMapping
import org.mapstruct.Mapper
import org.mapstruct.ReportingPolicy
import org.mapstruct.factory.Mappers

data class RecursiveTreeInput(
    val name: String,
    val childNodes: List<RecursiveTreeInput>?
): Input<TreeNode> {

    override fun toEntity(): TreeNode =
        CONVERTER.toTreeNode(this)

    @Mapper
    internal interface Converter {

        @BeanMapping(unmappedTargetPolicy = ReportingPolicy.IGNORE)
        fun toTreeNode(input: RecursiveTreeInput): TreeNode
    }

    companion object {
        @JvmStatic
        private val CONVERTER = Mappers.getMapper(Converter::class.java)
    }
}